back-end/friend.resolver.js

```
@@ -0,0 +1,177 @@
+    },
+    acceptFriendRequest: async (obj, { nickname }, { Users, Friends, req }) => {
+      const idFromNickname = await Users.findOne({
+        where: { nickname: nickname },
+      });
+      try {
+        await Friends.create({
+          pFriendId: idFromNickname.dataValues.id,
+          sFriendId: req.user.id,
+        });
+        await Friends.create({
+          pFriendId: req.user.id,
+          sFriendId: idFromNickname.dataValues.id,
+        });
+      } catch {
+        console.log('already exists');
+      }
+      await Friends.update(
+        { friendStateId: 2 },
+        {
+          where: {
+            [Op.and]: [{ pFriendId: idFromNickname.dataValues.id }, { sFriendId: req.user.id }],
+          },
+        },
+      );
+      return {
+        user: idFromNickname,
+        result: true,
+      };
+    },
+
+    sendFriendRequest: async (obj, { nickname }, { Friends, BeforeFriends, Users, req }) => {
+      // 이미 친구인지 확인
+      const checkFriendsTable = await Friends.findAll({
+        include: [
+          {
+            model: Users,
+            as: 'sFriend',
+            where: { [Op.or]: [{ nickname: nickname }, { id: req.user.id }] },
+          },
+          {
+            model: Users,
+            as: 'pFriend',
+            where: { [Op.or]: [{ nickname: nickname }, { id: req.user.id }] },
+          },
+        ],
+      });
+
+      if (checkFriendsTable.length > 1) throw new Error(`${nickname}님과는 이미 친구입니다.`);
+
+      // 친구 요청
+      const idFromNickname = await Users.findOne({
+        where: { nickname: nickname },
+      });
+
+      try {
+        await BeforeFriends.create({
+          pFriendId: req.user.id,
+          sFriendId: idFromNickname.dataValues.id,
+          friendStateId: 1,
+        });
+      } catch (e) {
+        if (e.message === 'Validation error') {
+          throw new Error(`이미 친구 신청을 보냈습니다.`);
+        }
+        throw new Error(`${nickname}님은 유저가 아닙니다.`);
```

일반적으로 파라미터로 전달 받은 key에 해당하는 row가 DB에 존재하는지 아닌지 체크하는 것은 로직의 최상단에 있는 것이 자연스럽기도 하고 fail fast에도 용이할 것 같습니다.

---

와 진짜 섬세하게 알려주셔서 감사합니다! 궁금한것도 해결되고, 리뷰 내용도 많은 도움 되었습니다. 👀 감사드려요!!